qcodes/tests/dataset/test_experiment_container.py

```diff
@@ -306,3 +312,33 @@ def test_load_last_experiment(empty_temp_db):
     assert last_exp.exp_id == exp2.exp_id
     assert last_exp.exp_id != exp1.exp_id
     assert last_exp.path_to_db == exp2.path_to_db
+
+
+@pytest.mark.usefixtures("empty_temp_db")
+def test_active_experiment(empty_temp_db_connection):
+    with pytest.raises(ValueError):
+        get_default_experiment_id(empty_temp_db_connection)
+
+    exp_1 = load_or_create_experiment("test_exp", sample_name="no_sample")
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_1.exp_id
+
+    exp_2 = new_experiment("test_exp_2", sample_name="no_sample")
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_2.exp_id
+
+    exp_3 = load_experiment(1)
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_1.exp_id
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_3.exp_id
+
+    exp_4 = new_experiment("test_exp_3", sample_name="no_sample")
+
+    exp_5 = load_experiment_by_name("test_exp_2", sample="no_sample")
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_2.exp_id
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_5.exp_id
+
+    exp_6 = load_last_experiment()
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_4.exp_id
+    assert get_default_experiment_id(empty_temp_db_connection) == exp_6.exp_id
+
+    reset_default_experiment_id()
+    with pytest.raises(ValueError):
```

Comment:
This should not happen. It should return the last experiment. However as you are using the connection fixture and not passing the connection to new experiment the experiment is not actually created in this db but in the default db of the user. 

Thinking a bit more of this I think we need a per dbpath default experiment for this to be safe

Reply:
I was initially thinking that the reset function should not be used publically, but in a very certain case, it could be used.
I think this can be fixed by adding a connection argument to the reset function:

```
def reset_default_experiment_id(conn) -> None:
    """
    Resets the default_experiment to the connected database default.
    """
    global default_experiment
    default_experiment = get_last_experiment(conn)
```

And we should change the test to create a temp database

Reply:
I am not convinced that is sufficient. I still think it would be better to index the active experiment by db path

Reply:
No, It is not. you are right